components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/impl/MLModelHandler.java

```diff
@@ -691,52 +714,84 @@ public MLModel retrieveModel(long modelId) throws MLModelHandlerException {
-
-        } catch (DatabaseHandlerException e) {
-            throw new MLModelPublisherException(errorMsg, e);
-        } catch (MLInputAdapterException e) {
-            throw new MLModelPublisherException(errorMsg, e);
-        } catch (MLOutputAdapterException e) {
-            throw new MLModelPublisherException(errorMsg, e);
-        } finally {
-            if (in != null) {
-                try {
-                    in.close();
-                } catch (IOException ignore) {
+        switch (mode) {
+        case SERIALIZED:
+            try {
+                // read model
+                MLStorage storage = databaseService.getModelStorage(modelId);
+                if (storage == null) {
+                    throw new InvalidRequestException("Invalid model [id] " + modelId);
+                }
+                String storageType = storage.getType();
+                String storageLocation = storage.getLocation();
+                MLIOFactory ioFactory = new MLIOFactory(mlProperties);
+                MLInputAdapter inputAdapter = ioFactory.getInputAdapter(storageType + MLConstants.IN_SUFFIX);
+                in = inputAdapter.read(storageLocation);
+                if (in == null) {
+                    throw new InvalidRequestException("Invalid model [id] " + modelId);
+                }
+                // create registry path
+                MLCoreServiceValueHolder valueHolder = MLCoreServiceValueHolder.getInstance();
+                String modelName = databaseService.getModel(tenantId, userName, modelId).getName();
+                relativeRegistryPath = "/" + valueHolder.getModelRegistryLocation() + "/" + modelName;
+                // publish to registry
+                registryOutputAdapter.write(relativeRegistryPath, in);
+            } catch (DatabaseHandlerException e) {
+                throw new MLModelPublisherException(errorMsg, e);
+            } catch (MLInputAdapterException e) {
+                throw new MLModelPublisherException(errorMsg, e);
+            } catch (MLOutputAdapterException e) {
+                throw new MLModelPublisherException(errorMsg, e);
+            } finally {
+                if (in != null) {
+                    try {
+                        in.close();
+                    } catch (IOException ignore) {
+                    }
                 }
             }
+            break;
+
+        case PMML:
+            MLCoreServiceValueHolder valueHolder = MLCoreServiceValueHolder.getInstance();
+            try {
+                String modelName = databaseService.getModel(tenantId, userName, modelId).getName();
+                relativeRegistryPath = "/" + valueHolder.getModelRegistryLocation() + "/" + modelName + ".xml";
+
+                MLModel model = retrieveModel(modelId);
+                String pmmlModel = exportAsPMML(model);
+                InputStream stream = new ByteArrayInputStream(pmmlModel.getBytes(StandardCharsets.UTF_8));
```

Comment:
Can use the InputStream "in" defined at the start of the method. No need to define a new input stream. Close the stream in finally block.